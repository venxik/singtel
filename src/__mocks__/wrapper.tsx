import { Provider } from 'react-redux';
import { AppStore, RootState, setupStore } from '../redux';
import { PreloadedState } from '@reduxjs/toolkit';
import { PropsWithChildren } from 'react';
import { render } from '@testing-library/react-native';

interface ExtendedRenderOptions {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}

export function AllTheProviders({ children }: { children: JSX.Element }) {
  return <Provider store={setupStore()}>{children}</Provider>;
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return <Provider store={store}>{children}</Provider>;
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}
