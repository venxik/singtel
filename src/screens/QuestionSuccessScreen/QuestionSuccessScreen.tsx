import { StyleSheet } from 'react-native';

import { Button, ComicText, View } from '../../components';
import { useEffect } from 'react';
import { RouteProp, useNavigation, useRoute } from '@react-navigation/native';
import { RootStackNavigationProps, RootStackParamList } from '../../navigation';
import { useReduxDispatch, useReduxSelector } from '../../redux';
import {
  addCurrentScore,
  addScoreToLeaderboard,
  resetCurrentScore,
  resetQuestionTracker,
} from '../../redux/main';

type QuestionSuccessScreenRouteProp = RouteProp<RootStackParamList, 'QuestionSuccessScreen'>;

const QuestionSuccessScreen = () => {
  const dispatch = useReduxDispatch();
  const { questionTracker, currentScore } = useReduxSelector((s) => s.main);

  const { replace } = useNavigation<RootStackNavigationProps<'QuestionSuccessScreen'>>();
  const route = useRoute<QuestionSuccessScreenRouteProp>();

  const { points } = route.params;

  useEffect(() => {
    dispatch(addCurrentScore(points));
  }, []);

  const onPressNext = () => {
    if (questionTracker === 3) {
      dispatch(addScoreToLeaderboard(currentScore));
      dispatch(resetCurrentScore());
      dispatch(resetQuestionTracker());
      replace('LeaderboardScreen');
      return;
    }
    replace('QuestionScreen');
  };

  return (
    <View style={styles.container}>
      <ComicText
        style={styles.title}
        testID="title-text">{`Correct! Congratulations You Earn ${points} Points`}</ComicText>
      <Button testID={`button-next`} style={styles.button} onPress={onPressNext}>
        <ComicText>Next</ComicText>
      </Button>
    </View>
  );
};

export default QuestionSuccessScreen;

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
  },
  title: { fontSize: 30, marginVertical: '20%', textAlign: 'center', paddingHorizontal: 40 },
  button: {
    position: 'absolute',
    bottom: 20,
    width: '90%',
  },
});
