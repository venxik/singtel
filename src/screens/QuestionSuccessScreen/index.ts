import QuestionSuccessScreen from './QuestionSuccessScreen';

export { QuestionSuccessScreen };
