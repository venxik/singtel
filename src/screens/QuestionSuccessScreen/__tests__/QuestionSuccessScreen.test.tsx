import { fireEvent, screen } from '@testing-library/react-native';
import { renderWithProviders } from '../../../__mocks__';
import QuestionSuccessScreen from '../QuestionSuccessScreen';

const mockedNavigate = jest.fn();

jest.mock('@react-navigation/native', () => {
  const actualNav = jest.requireActual('@react-navigation/native');
  return {
    ...actualNav,
    useNavigation: () => ({
      replace: mockedNavigate,
    }),
    useRoute: () => ({
      params: {
        points: 10,
      },
    }),
  };
});

describe('QuestionSuccessScreen', () => {
  it('match snapshot', async () => {
    renderWithProviders(<QuestionSuccessScreen />);
    expect(screen.toJSON()).toMatchSnapshot();
  });

  it('have correct point values', async () => {
    renderWithProviders(<QuestionSuccessScreen />);
    expect(screen.getByTestId('title-text')).toHaveTextContent(
      'Correct! Congratulations You Earn 10 Points'
    );
  });

  it('can navigate to question screen when questionTracker below 3', async () => {
    renderWithProviders(<QuestionSuccessScreen />);
    fireEvent.press(screen.getByTestId('button-next'));

    expect(mockedNavigate).toHaveBeenCalledWith('QuestionScreen');
  });

  it('can navigate to leaderboard screen when questionTracker is 3 or finished', async () => {
    renderWithProviders(<QuestionSuccessScreen />, {
      preloadedState: {
        main: {
          currentScore: 0,
          leaderboards: [],
          questionTracker: 3,
          selectedCategory: '',
        },
      },
    });
    fireEvent.press(screen.getByTestId('button-next'));

    expect(mockedNavigate).toHaveBeenCalledWith('LeaderboardScreen');
  });
});
