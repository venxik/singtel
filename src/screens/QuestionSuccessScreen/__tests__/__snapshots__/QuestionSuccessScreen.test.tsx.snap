// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LeaderboardScreen match snapshot 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#fff",
      },
      {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      },
    ]
  }
>
  <Text
    style={
      [
        {
          "color": "#000",
        },
        [
          {
            "fontSize": 30,
            "marginVertical": "20%",
            "paddingHorizontal": 40,
            "textAlign": "center",
          },
          {
            "fontFamily": "comic",
          },
        ],
      ]
    }
    testID="title-text"
  >
    Correct! Congratulations You Earn 10 Points
  </Text>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#686fcc",
        "borderRadius": 20,
        "bottom": 20,
        "opacity": 1,
        "paddingHorizontal": "15%",
        "paddingVertical": "5%",
        "position": "absolute",
        "width": "90%",
      }
    }
    testID="button-next"
  >
    <Text
      style={
        [
          {
            "color": "#000",
          },
          [
            undefined,
            {
              "fontFamily": "comic",
            },
          ],
        ]
      }
    >
      Next
    </Text>
  </View>
</View>
`;

exports[`QuestionSuccessScreen match snapshot 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#fff",
      },
      {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      },
    ]
  }
>
  <Text
    style={
      [
        {
          "color": "#000",
        },
        [
          {
            "fontSize": 30,
            "marginVertical": "20%",
            "paddingHorizontal": 40,
            "textAlign": "center",
          },
          {
            "fontFamily": "comic",
          },
        ],
      ]
    }
    testID="title-text"
  >
    Correct! Congratulations You Earn 10 Points
  </Text>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#686fcc",
        "borderRadius": 20,
        "bottom": 20,
        "opacity": 1,
        "paddingHorizontal": "15%",
        "paddingVertical": "5%",
        "position": "absolute",
        "width": "90%",
      }
    }
    testID="button-next"
  >
    <Text
      style={
        [
          {
            "color": "#000",
          },
          [
            undefined,
            {
              "fontFamily": "comic",
            },
          ],
        ]
      }
    >
      Next
    </Text>
  </View>
</View>
`;
